federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -0,0 +1,830 @@
+}
+
+func (s *ServiceController)updateAllServicesToCluster(services []*cachedService, clusterName string){
+	cluster, ok := s.clusterCache.clientMap[clusterName]
+	if ok {
+		for _, cachedService := range services{
+			appliedState := cachedService.appliedState
+			s.processServiceForCluster(clusterName, appliedState, cluster.clientset)
+		}
+	}
+}
+
+func (s *ServiceController)removeAllServicesFromCluster(services []*cachedService, clusterName string){
+	client, ok := s.clusterCache.clientMap[clusterName]
+	if ok {
+		for _, cachedService := range services{
+			appliedState := cachedService.appliedState
+			s.deleteClusterService(clusterName, appliedState, client.clientset)
+		}
+		glog.Infof("synced all services to cluster %s", clusterName)
+	}
+}
+
+// updateDNSRecords updates all existing federated service DNS Records so that
+// they will match the list of host names provided.
+// Returns the list of services that couldn't be updated.
+func (s *ServiceController) updateDNSRecords(services []*cachedService, clusters []string) (servicesToRetry []*cachedService) {
+	for _, service := range services {
+		func() {
+			service.rwlock.Lock()
+			defer service.rwlock.Unlock()
+			// If the applied state is nil, that means it hasn't yet been successfully dealt
+			// with by the DNS Record reconciler. We can trust the DNS Record
+			// reconciler to ensure the federated service's DNS records are created to target
+			// the correct backend service IP's
+			if service.appliedState == nil {
+				return
+			}
+			if err := s.lockedUpdateDNSRecords(service, clusters); err != nil {
+				glog.Errorf("External error while updating DNS Records: %v.", err)
+				servicesToRetry = append(servicesToRetry, service)
+			}
+		}()
+	}
+	return servicesToRetry
+}
+
+// Updates the DNS records of a service, assuming we hold the mutex
+// associated with the service.
+// TODO: quinton: Still screwed up in the same way as above.  Fix.
+func (s *ServiceController) lockedUpdateDNSRecords(service *cachedService, clusterNames []string) error {
+	if !wantsDNSRecords(service.appliedState) {
+		return nil
+	}
+	for key, _ := range s.clusterCache.clientMap {
+		for _, clusterName := range clusterNames {
+			if key == clusterName {
+				ensureDNSRecords(clusterName, service)
+			}
+		}
+	}
+	return nil
+}
+
+func LoadBalancerIPsAreEqual(oldService, newService *v1.Service) bool {
+	return oldService.Spec.LoadBalancerIP == newService.Spec.LoadBalancerIP
+}
+
+// Computes the next retry, using exponential backoff
+// mutex must be held.
+func (s *cachedService) nextRetryDelay() time.Duration {
+	s.lastRetryDelay = s.lastRetryDelay * 2
+	if s.lastRetryDelay < minRetryDelay {
+		s.lastRetryDelay = minRetryDelay
+	}
+	if s.lastRetryDelay > maxRetryDelay {
+		s.lastRetryDelay = maxRetryDelay
+	}
+	return s.lastRetryDelay
+}
+
+// Resets the retry exponential backoff.  mutex must be held.
+func (s *cachedService) resetRetryDelay() {
+	s.lastRetryDelay = time.Duration(0)
```

Comment:
I assume you need to reset to non-zero, as exponentially increasing zero still yields zero?